File: Array/Coin_Change/Coin_Change_II.py
class Solution:
        def change(self, amount: int, coins: List[int]) -> int:
                if amount <= 0:
                        return 1
                if len(coins) == 0:
                        return 0  
                
                cache = {}
                coins.sort()
                
                #currentsum, index, instruction
                stack = [[0, 0, 'ENTER']]
                
                while stack:
                        currentsum, index, instruction = stack.pop()
                        if currentsum == amount:
                                cache[(currentsum, index)] = 1
                        elif instruction == 'ENTER':
                                stack.append([currentsum, index, 'EXIT'])
                        
                                if (index+1) < len(coins) and (currentsum, index+1) not in cache.keys():
                                        stack.append([currentsum, index+1, 'ENTER'])
                        
                                newsum = currentsum + coins[index]
                                if newsum <= amount and (newsum, index) not in cache.keys():
                                        stack.append([newsum, index, 'ENTER'])
                        elif instruction == 'EXIT':
                                newsum = currentsum + coins[index]
                                left = cache.get((newsum, index)) or 0
                                right = cache.get((currentsum, index+1)) or 0
                                cache[(currentsum, index)] = left + right
                
                return cache[(0,0)]

File: Array/Coin_Change/Coin_Change_II.ts
// Leetcode: https://leetcode.com/problems/coin-change-ii/

// SOLUTION 1: RECURSION (TLE)
var change = function (amount: number, coins: number[]): number {
    let count: number = 0;
    coins.sort((a, b) => a - b);

    const combinationSum = (sum: number = 0, idx: number = 0): number => {
        if (sum < 0 || sum > amount) {
            return 0;
        }

        if (sum == amount) {
            count = count + 1;
            return count;
        }

        for (let i: number = idx; i < coins.length; i++) {
            combinationSum(sum + coins[i], i);
        }
    };

    combinationSum();
    return count;
};

// SOLUTION 2: RECURSION WITH MEMOIZATION
var change = function (amount: number, coins: number[]): number {
    let cache: {
        [key: string]: number;
    } = {};
    coins.sort((a, b) => a - b);

    const combinationSum = (sum: number = 0, idx: number = 0): number => {
        let count: number = 0;
        if (sum < 0 || sum > amount) {
            return count;
        }

        const key: string = `${sum},${idx}`;
        if (key in cache) {
            count += cache[key];
            return count;
        }

        if (sum === amount) {
            count = count + 1;
            return count;
        }

        for (let i: number = idx; i < coins.length; i++) {
            count += combinationSum(sum + coins[i], i);
        }
        cache[key] = count;
        return count;
    };

    return combinationSum();
};
